src/lib/es2015.promise.d.ts

```diff
@@ -10,7 +10,7 @@ interface PromiseConstructor {
      * a resolve callback used resolve the promise with a value or the result of another promise,
      * and a reject callback used to reject the promise with a provided reason or error.
      */
-    new <T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
+    new <T>(executor: (resolve: [T] extends [void] ? (value?: T | PromiseLike<T>) => void : (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;

```

Comment:
Is the goal to only allow `value` to be optional if `T` is exactly `void`? If `T` is `void | number` you end up with the second definition (with a non-optional `value`) for `(value: void | number | PromiseLike<void | number>) => void`.

You could try this instead:
```ts
resolve: [void] extends [T] ? (value?: T | PromiseLike<T>) => void : (value: T | PromiseLike<T>) => void
```

By inverting the condition, you would get cases where `void` is assignable (such as when `T` is a union including `void`).

Reply:
It seems a nice idea. I'll do it.

Reply:
Done.

Reply:
I realized `[void] extends [T]` doesn't work with `undefined`. So I reverted it.

Reply:
That's unfortunate, but its better than nothing I suppose.

Reply:
Have you tried this without the conditional type? This should work as well and is probably easier to read:

```ts
    new <T extends void>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
    new <T>(executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;

```

Reply:
As I mentioned at https://github.com/Microsoft/TypeScript/pull/21867#issuecomment-375087677, it is not extensible.